1. Basics/task_7/part_4/task_4_on_apache_handler.php

```diff
@@ -0,0 +1,188 @@
+		echo '<h1 style=\'color:green\'>FOUND</h1>';
+    } catch(Exception $ex) {
+		http_response_code($ex->getCode());
+        echo '<p>' . $ex->getMessage() . '</p>';
+    };	
+}
+
+// REQUEST PROCESS FUNCTIONS. VALIDATON FUNCTIONS /////////////////////////////////////////////////
+/**
+ * @throws Exception
+ * @return void
+ */
+function checkRequestMethod() : void {
+	if(($_SERVER['REQUEST_METHOD'] ?? null) === 'POST') {
+		return;
+	}
+
+    throw new Exception('Request method is invalid. The method POST is needed.', 400);
+}
+
+/**
+ * @throws Exception
+ * @return void
+ */
+function checkRequestUri() : void {		
+	if(preg_match('#^/api/checkLoginAndPassword$#', $_SERVER['REQUEST_URI'] ?? '')) {
+		return;
+	}
+
+    throw new Exception('URI is invalid.', 404);
+}
+
+/**
+ * @throws Exception
+ * @return void
+ */
+function checkRequestContentTypeValue() : void {
+   	if(preg_match('#\bapplication/x-www-form-urlencoded\b#i', apache_request_headers()['Content-Type'] ?? '')) {
+		return;
+	}
+    
+    throw new Exception('Header\'s \'Content-Type\' value is invalid.', 400);
+}
+
+/**
+ * @throws Exception
+ * @return array
+ */
+function validateLoginAndPassword(): array {
+	$login = sanitizeInput($_POST['login'] ?? false);
+
+	if($login === false) {
+		throw new Exception('Login field is absent', 404);
+	}
+
+	if(empty($login)) {
+		throw new Exception('Login value is absent(empty field)', 404);
+	}
+
+	$password = sanitizeInput($_POST['password'] ?? false);
+	
+	if($password === false) {
+		throw new Exception('Password field is absent', 404);
+	}
+	
+	if(empty($password)) {
+		throw new Exception('Password value is absent(empty field)', 404);
+	}
+
+	return ['login' => $login, 'password' => $password,];
+}
+
+/**
+ * @param string $fileName
+ * @throws Exception
+ * @return void
+ */
+function checkFileExistance(string $fileName) : void {
+    if(file_exists($fileName)) {
+		return;
+	}
+    
+    throw new Exception('File ' . $fileName . ' doesn`t exist', 500);
+}
+
+/**
+ * @param string $login
+ * @param string $password
+ * @throws Exception
+ * @return void
+ */
+function validateUser(string $login, string $password) : void {
+    $logAndPassPairsAsArr = getFileContentAsPairsArray(FILE_NAME, PHP_EOL, ':');
+
+    foreach($logAndPassPairsAsArr as $log => $pass) {
+        if($log === $login and $pass === $password) {
+			return;
+		}
+              
+		if($log === $login) {
+            throw new Exception('Password didn`t match. Fogot password?', 404);        
+        }
+    }
+
+    throw new Exception('User not found. Would you like to register?', 404);
+}
+
+// FILES PROCESSING FUNCTIONS ////////////////////////////////////////////////////
+/**
+ * @param string $filename
+ * @param string $delim1
+ * @param string $delim2
+ * 
+ * @return array
+ */
+function getFileContentAsPairsArray($filename, $delim1, $delim2) : array {
+    return getStringContentAsPairsArray(file_get_contents($filename), $delim1, $delim2);
```

Comment:
`file_get_contents` is a bad idea for reading files, especially with ones which will grow with time. do it via file streams